### PR TITLE
Add API to add a memory buffer to an index

### PR DESCRIPTION
**Pull request by dprofeta**
git_index_add_frombuffer enables now to store a memory buffer in the odb
and to store is in the index directly, even on a bare index.

### CODE REVIEW

**Comment by dprofeta** [comment]
This is should fix the issue https://github.com/libgit2/libgit2/issues/2876

File: tests/index/tests.c [hunk 1]
```diff
@@ -253,6 +253,56 @@ void test_index_tests__add(void)
...
+		"hey there\n", 10));
+
+	/* Wow... it worked! */
+	cl_assert(git_index_entrycount(index) == 1);
+	entry = git_index_get_byindex(index, 0);
+
+	/* And the built-in hashing mechanism worked as expected */
+	cl_assert_equal_oid(&id1, &entry->id);
+	/* And mode is the one asked */
+	cl_assert(entry->mode == GIT_FILEMODE_BLOB);
```
Comment (carlosmn) [inline tests/index/tests.c]: Let's use `cl_assert_equal_i(GIT_FILEMODE_BLOB, entry->mode)`. This way clar will report the values.

**Comment by carlosmn** [comment]
We should test (and document) what happens with the other data in the index entry. We should also test what happens when we write to an entry which already exists, as there the rest of the data is even more important.

Should we zero out all of the other data? We should test and mention that. If we don't change the stat data, performing a status would likely do something weird. If we do, then we need to decide in which direction, and we need to test for that.

**Comment by dprofeta** [comment]
I don't see why we should touch the other data in the index entry. The behaviour should be consistent with the current git_index_add_bypath, that's just another way to add data in the index. 
The usecase I see if for a bare repo to create a new commit on the fly thanks to a in-memory index. So there is no status that make sense. Your question is when we use the method with a full repo? The status will show the file added in the index and deleted in the working directory. I can add a test and document that if it make sense for you.

**Comment by carlosmn** [comment]
>I don't see why we should touch the other data in the index entry. The behaviour should be consistent with the current git_index_add_bypath, that's just another way to add data in the index. 

We are not doing the same thing. `git_index_add_bypath()` fills in the stat data from disk. We do not do this in this function. Currently it does not change the stat data we get from the allocator, which means it invalidates it. This is fine, but it's implementation-defined, which is bad because if this changes, it can affect status and diff.

> The status will show the file added in the index and deleted in the working directory.

It will if it's a new file; but if you update an entry for an existing file, it could show that it's changed, or it could still think that it has the data which is on the filesystem and skip looking at it, depending on what happens to the stat information. As we do not say what happens, it's not possible to rely on this function to act on an index which will be written to disk.

**Comment by dprofeta** [comment]
Oh, ok I see the issue now.
We can indeed zero the other entry but it may not be the best thing to do.
Instead of the `git_filemode_t` in parameter of the method, we can have a `struct stat`, so that the user can put the data he wants. Maybe with a helper function to extract it from the previous entry.
Or we can update the best we can the information. We would zero the ino and dev, update the mtime and the size. But then for uid and gid, I don't really know what we can do…. 
Do you have any recommendation?

**Comment by ethomson** [comment]
I think a `struct stat` is probably overkill, and possibly confusing.  My naive assumption is that we should zero it.

**Comment by nulltoken** [comment]
@ethomson That sounds like a good idea.

**Comment by carlosmn** [comment]
Zeroing (thus invalidating) sounds reasonable (but we have to document and test that). We also already have a way of getting all this information from a user, namely `git_index_entry`. This function isn't a variant of `git_index_add_bypath()` but of `git_index_add()`, thus I think it makes more sense to ask the user for a `git_index_entry` which we use just like in `git_index_add()` with the exception that we override the id with whatever the buffer happens to be. This will still invalidate in the usual case and would still let the caller set whichever fields they want.

**Comment by ethomson** [comment]
Hmm, I was thinking that a `git_index_entry` might be a bit of overkill, but the more I think about it, the more I like it.  I assume this is optional, and if I pass `NULL`, I'll get some reasonable defaults?  I like that...

**Comment by carlosmn** [comment]
I was thinking of making it compulsory and making you pass in the path via the index entry as well, so we behave just like `git_index_add()` but do an extra step. I suppose we could compromise and let you pass in a path and then override the path in the entry you pass in as well.

**Comment by ethomson** [comment]
Ah.  Yeah, the path...  I figured we'd be overriding the ID and the size for sure...  The path does seem to get duplicated up a bit.

**Comment by dprofeta** [comment]
Building a `git_index_entry` is not easy today. I would like to put a function like `index_entry_create()` in the public api to ease the creation with the right path and then asking for it in the `git_index_add_frombuffer()` would be easier.

**Comment by ethomson** [comment]
I was pondering this a bit and I'm struggling to think of a case where I want to both provide a buffer _and_ all the `git_index_entry` information...  I feel like this is probably not backed by a file (else the user would want `git_index_add_bypath`), so the `struct stat`-like information seems unnecessary...  I don't think this is appropriate for adding high-stage index entries...

Is there a use case for providing a buffer and `git_index_entry` both that is compelling?

**Comment by ethomson** [comment]
Also, @shiftkey was thinking about this for LibGit2Sharp, if I recall?

**Comment by carlosmn** [comment]
Creating an `git_index_entry` is trivial. It is a public struct, which you create on the stack or the heap just like any other struct. Any function which we provide will only hinder, as we have no idea which allocator the user wants.

>Is there a use case for providing a buffer and git_index_entry both that is compelling?

Hand-rolled filtering. But the main reason I think doing it this way makes sense is that this is the way we have now, and this is a convenience function which does one of the steps for you.

In any case, I have said I'm fine with zeroing everything else out (which is what most people would do anyway), but we need to test and document that behaviour.

**Comment by shiftkey** [comment]
> Also, @shiftkey was thinking about this for LibGit2Sharp, if I recall?

Uhhhh, maybe? :eyes:-ing

**Comment by nulltoken** [comment]
>> Also, @shiftkey was thinking about this for LibGit2Sharp, if I recall?

>Uhhhh, maybe? :eyes:-ing 

FWIW, one can already achieve this today through `ObjectDatabase.CreateBlob((Memory)Stream)` and `Index.Add(Blob)` method.

File: CHANGELOG.md [hunk 2]
```diff
@@ -14,6 +14,10 @@ v0.22 + 1
   via `git_config_parse_path()` and `git_config_get_path()`
   respectively.
 
+* `git_index_add_frombuffer()` can now create a blob from memory
+   buffer and add it to the index, even if the index is also in
+   memory.
+
```
Comment (carlosmn) [inline CHANGELOG.md]: This does not sound accurate. If the index in purely in-memory, we have no associated repository and thus cannot know where to create the blob (there is even a check for this in the function).

Comment (dprofeta) [inline CHANGELOG.md]: You can create an in-memory index and associate a repository to it with `git_repository_set_index`. Maybe, I can add that the index need to be associtated to a repository.

**Comment by dprofeta** [comment]
Is there anything left to update here or is the pull request fine now?

File: include/git2/index.h [hunk 3]
```diff
@@ -457,6 +457,38 @@ GIT_EXTERN(int) git_index_entry_stage(const git_index_entry *entry);
 GIT_EXTERN(int) git_index_add_bypath(git_index *index, const char *path);
 
 /**
+ * Add or update an index entry from a buffer in memory
+ *
+ * This method will create a blob in the repository that owns the
+ * index and then add the blob to the index.  The `path` of the entry
+ * represents the position of the blob relative to the repository's
+ * root folder.
```
Comment (carlosmn) [inline include/git2/index.h]: This wording seems off. You don't add a blob to the index, you add an entry. The blob is also pure content. It does not have a path at all.

File: CHANGELOG.md [hunk 4]
```diff
@@ -14,6 +14,10 @@ v0.22 + 1
   via `git_config_parse_path()` and `git_config_get_path()`
   respectively.
 
+* `git_index_add_frombuffer()` can now create a blob from memory
+   buffer and add it to the index, even if the index is also in
+   memory. The index must be attach to a repository to store the blob.
```
Comment (carlosmn) [inline CHANGELOG.md]: This is contradictory. If the index is attached to a repository, it is not a purely in-memory index. I would just go with

>...add it to an index which is attached to a repository.

File: tests/index/tests.c [hunk 5]
```diff
@@ -253,6 +253,125 @@ void test_index_tests__add(void)
...
+	cl_git_pass(git_oid_fromstr(&id1, "a8233120f6ad708f843d861ce2b7228ec4e3dec6"));
+
+	cl_git_pass(git_index_add_bypath(index, "test.txt"));
+
+	/* Add the new file to the index */
+	memset(&entry, 0x0, sizeof(git_index_entry));
+	entry.mode = GIT_FILEMODE_BLOB;
+	entry.path = "test.txt";
+	cl_git_pass(git_index_add_frombuffer(index, &entry,
+		"hey there\n", 10));
```
Comment (carlosmn) [inline tests/index/tests.c]: Let's make the string a const and have the compiler give us its length via `strlen(content)`, here and below so as to reduce the chance of error.

File: src/index.c [hunk 6]
```diff
@@ -1097,6 +1097,58 @@ static int index_conflict_to_reuc(git_index *index, const char *path)
...
+	if (index_entry_dup(&entry, INDEX_OWNER(index), source_entry) < 0)
+		return -1;
+
+	error = git_blob_create_frombuffer(&id, INDEX_OWNER(index), buffer, len);
+	if (error < 0) {
+		index_entry_free(entry);
+		return error;
+	}
+
+	entry->id = id;
```
Comment (carlosmn) [inline src/index.c]: We make explicit use of `git_oid_cpy()` for copying `git_oid`s around.

**Comment by carlosmn** [comment]
:+1: @ethomson ?

**Comment by ethomson** [comment]
:+1: Manually merged.  Thanks!

**Pull request closed without merging**